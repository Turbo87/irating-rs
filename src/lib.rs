use std::f32::consts::LN_2;

#[derive(Debug, Clone)]
pub struct RaceResult<D = ()> {
    pub driver: D,
    pub finish_rank: u32,
    pub start_irating: u32,
    pub started: bool,
}

impl<D> From<(D, u32, u32, bool)> for RaceResult<D> {
    fn from(value: (D, u32, u32, bool)) -> Self {
        RaceResult {
            driver: value.0,
            finish_rank: value.1,
            start_irating: value.2,
            started: value.3,
        }
    }
}

#[derive(Debug, Clone)]
pub struct CalculationResult<D = ()> {
    pub race_result: RaceResult<D>,
    pub irating_change: f32,
    pub new_irating: u32,
}

pub fn calculate<D>(race_results: Vec<RaceResult<D>>) -> Vec<CalculationResult<D>> {
    let br1 = 1600. / LN_2;

    let num_registrations = race_results.len();
    let num_starters = race_results.iter().filter(|result| result.started).count();
    let num_non_starters = num_registrations - num_starters;

    let chances: Vec<Vec<f32>> = race_results
        .iter()
        .map(|result| result.start_irating as f32)
        .map(|a| {
            race_results
                .iter()
                .map(|result| result.start_irating as f32)
                .map(|b| chance(a, b, br1))
                .collect()
        })
        .collect();

    // this appears to be unused?
    //
    // let sof_exponential: Vec<f32> = race_results
    //     .iter()
    //     .map(|result| result.start_irating as f32)
    //     .map(|irating| (-irating / br1).exp())
    //     .collect();

    let expected_scores: Vec<f32> = chances
        .iter()
        .map(|chances| chances.iter().sum::<f32>() - 0.5)
        .collect();

    let fudge_factors: Vec<f32> = race_results
        .iter()
        .map(|result| match result.started {
            false => 0.,
            true => {
                let x = num_registrations as f32 - num_non_starters as f32 / 2.;
                (x / 2. - result.finish_rank as f32) / 100.
            }
        })
        .collect();

    let changes_starters: Vec<Option<f32>> = race_results
        .iter()
        .zip(expected_scores.iter())
        .zip(fudge_factors.iter())
        .map(
            |((result, expected_score), fudge_factor)| match result.started {
                false => None,
                true => Some(
                    (num_registrations as f32
                        - result.finish_rank as f32
                        - expected_score
                        - fudge_factor)
                        * 200.
                        / num_starters as f32,
                ),
            },
        )
        .collect();

    let sum_changes_starters: f32 = changes_starters.iter().filter_map(Option::as_ref).sum();

    let expected_score_non_starters: Vec<Option<f32>> = race_results
        .iter()
        .zip(expected_scores.iter())
        .map(|(result, expected_score)| (!result.started).then_some(*expected_score))
        .collect();

    let sum_expected_score_non_starters: f32 = expected_score_non_starters
        .iter()
        .filter_map(Option::as_ref)
        .sum();

    let changes_non_starters: Vec<Option<f32>> = expected_score_non_starters
        .iter()
        .map(|expected_score| {
            expected_score.map(|expected_score| {
                -sum_changes_starters / num_non_starters as f32 * expected_score
                    / (sum_expected_score_non_starters / num_non_starters as f32)
            })
            //
        })
        .collect();

    let changes: Vec<f32> = changes_starters
        .iter()
        .zip(changes_non_starters.iter())
        .map(|change| match change {
            (Some(change), None) => *change,
            (None, Some(change)) => *change,
            (_, _) => panic!(),
        })
        .collect();

    race_results
        .into_iter()
        .zip(changes.iter())
        .map(|(result, change)| {
            let new_irating = (result.start_irating as f32 + change).round() as u32;
            CalculationResult {
                race_result: result,
                irating_change: *change,
                new_irating,
            }
        })
        .collect()
}

fn chance(a: f32, b: f32, factor: f32) -> f32 {
    (1. - (-a / factor).exp()) * (-b / factor).exp()
        / ((1. - (-b / factor).exp()) * (-a / factor).exp()
            + (1. - (-a / factor).exp()) * (-b / factor).exp())
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
        let race_results = vec![
            ("Driver 1", 1, 7526, true).into(),
            ("Driver 2", 2, 5982, true).into(),
            ("Driver 3", 3, 5463, true).into(),
            ("Driver 4", 4, 4279, true).into(),
            ("Driver 5", 5, 4137, true).into(),
            ("Driver 6", 6, 4044, true).into(),
            ("Driver 7", 7, 3891, true).into(),
            ("Driver 8", 8, 3612, true).into(),
            ("Driver 9", 9, 3147, true).into(),
            ("Driver 10", 10, 2823, true).into(),
            ("Driver 11", 11, 2715, true).into(),
            ("Driver 12", 12, 2603, true).into(),
            ("Driver 13", 13, 2512, true).into(),
            ("Driver 14", 14, 2352, false).into(),
            ("Driver 15", 15, 2227, true).into(),
            ("Driver 16", 16, 2195, true).into(),
            ("Driver 17", 17, 2166, true).into(),
            ("Driver 18", 18, 2089, true).into(),
            ("Driver 19", 19, 1773, true).into(),
            ("Driver 20", 20, 1772, true).into(),
            ("Driver 21", 21, 1752, true).into(),
            ("Driver 22", 22, 1748, true).into(),
            ("Driver 23", 23, 1705, true).into(),
            ("Driver 24", 24, 1662, true).into(),
            ("Driver 25", 25, 1622, true).into(),
            ("Driver 26", 26, 1537, true).into(),
            ("Driver 27", 27, 1464, true).into(),
            ("Driver 28", 28, 1203, true).into(),
        ];

        let result = super::calculate(race_results);
        insta::assert_debug_snapshot!(result);
    }
}
