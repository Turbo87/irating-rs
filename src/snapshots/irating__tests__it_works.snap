---
source: src/lib.rs
expression: result
---
[
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 1",
            finish_rank: 1,
            start_irating: 7526,
            started: true,
        },
        irating_change: 17.63672,
        new_irating: 7544,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 2",
            finish_rank: 2,
            start_irating: 5982,
            started: true,
        },
        irating_change: 25.833923,
        new_irating: 6008,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 3",
            finish_rank: 3,
            start_irating: 5463,
            started: true,
        },
        irating_change: 25.432884,
        new_irating: 5488,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 4",
            finish_rank: 4,
            start_irating: 4279,
            started: true,
        },
        irating_change: 37.92791,
        new_irating: 4317,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 5",
            finish_rank: 5,
            start_irating: 4137,
            started: true,
        },
        irating_change: 33.394478,
        new_irating: 4170,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 6",
            finish_rank: 6,
            start_irating: 4044,
            started: true,
        },
        irating_change: 27.948332,
        new_irating: 4072,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 7",
            finish_rank: 7,
            start_irating: 3891,
            started: true,
        },
        irating_change: 23.814116,
        new_irating: 3915,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 8",
            finish_rank: 8,
            start_irating: 3612,
            started: true,
        },
        irating_change: 22.626814,
        new_irating: 3635,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 9",
            finish_rank: 9,
            start_irating: 3147,
            started: true,
        },
        irating_change: 26.485985,
        new_irating: 3173,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 10",
            finish_rank: 10,
            start_irating: 2823,
            started: true,
        },
        irating_change: 27.702335,
        new_irating: 2851,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 11",
            finish_rank: 11,
            start_irating: 2715,
            started: true,
        },
        irating_change: 23.36419,
        new_irating: 2738,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 12",
            finish_rank: 12,
            start_irating: 2603,
            started: true,
        },
        irating_change: 19.21653,
        new_irating: 2622,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 13",
            finish_rank: 13,
            start_irating: 2512,
            started: true,
        },
        irating_change: 14.53251,
        new_irating: 2527,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 14",
            finish_rank: 14,
            start_irating: 2352,
            started: false,
        },
        irating_change: 10.437519,
        new_irating: 2362,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 15",
            finish_rank: 15,
            start_irating: 2227,
            started: true,
        },
        irating_change: 8.5288105,
        new_irating: 2236,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 16",
            finish_rank: 16,
            start_irating: 2195,
            started: true,
        },
        irating_change: 2.2037997,
        new_irating: 2197,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 17",
            finish_rank: 17,
            start_irating: 2166,
            started: true,
        },
        irating_change: -4.2093577,
        new_irating: 2162,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 18",
            finish_rank: 18,
            start_irating: 2089,
            started: true,
        },
        irating_change: -9.06982,
        new_irating: 2080,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 19",
            finish_rank: 19,
            start_irating: 1773,
            started: true,
        },
        irating_change: -5.7882223,
        new_irating: 1767,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 20",
            finish_rank: 20,
            start_irating: 1772,
            started: true,
        },
        irating_change: -13.086736,
        new_irating: 1759,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 21",
            finish_rank: 21,
            start_irating: 1752,
            started: true,
        },
        irating_change: -19.722021,
        new_irating: 1732,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 22",
            finish_rank: 22,
            start_irating: 1748,
            started: true,
        },
        irating_change: -26.915356,
        new_irating: 1721,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 23",
            finish_rank: 23,
            start_irating: 1705,
            started: true,
        },
        irating_change: -32.73568,
        new_irating: 1672,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 24",
            finish_rank: 24,
            start_irating: 1662,
            started: true,
        },
        irating_change: -38.54108,
        new_irating: 1623,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 25",
            finish_rank: 25,
            start_irating: 1622,
            started: true,
        },
        irating_change: -44.439545,
        new_irating: 1578,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 26",
            finish_rank: 26,
            start_irating: 1537,
            started: true,
        },
        irating_change: -48.679874,
        new_irating: 1488,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 27",
            finish_rank: 27,
            start_irating: 1464,
            started: true,
        },
        irating_change: -53.308353,
        new_irating: 1411,
    },
    CalculationResult {
        race_result: RaceResult {
            driver: "Driver 28",
            finish_rank: 28,
            start_irating: 1203,
            started: true,
        },
        irating_change: -50.590836,
        new_irating: 1152,
    },
]
